src/models/ratings/ratings.service.ts

```diff
@@ -0,0 +1,492 @@
+		@InjectRepository(Games, 'games')
+		private gamesRepository: Repository<Games>,
+	) {}
+
+	async getAll(query?: RatingQuery): Promise<RatingList | DefaultExceptionDto> {
+		const limit = parseInt(query.limit) || 50;
+		const skip = parseInt(query.skip) || 0;
+		const page = parseInt(query.page) || 0;
+		const order: 'ASC' | 'DESC' = query.order || 'DESC';
+		const sort = query.sort ? query.sort : 'participation_rating';
+		const whereSearch = {
+			ratingbase: 0,
+			unrated: 0,
+			ratedgames: MoreThan(0),
+		}
+		if(query.name){ whereSearch['name'] = Like(`${query.name}`); }
+		if(query.id){ whereSearch['id'] = query.id; }
+		try {
+			const results = await this.ratingRepository.findAndCount({
+				select: ['name', 'rating', 'ratedgames', 'maxrating', 'participation_rating', 'isbot'],
+				where: whereSearch,
+				order: {
+					[sort]: order,
+				},
+				take: limit,
+				skip: limit * page || skip,
+			});
+			return {
+				items: results[0] || [],
+				total: results[1] || 0,
+				page: page + 1,
+				perPage: limit,
+				totalPages: Math.ceil(results[1] / limit),
+			};
+		} catch (error) {
+			this.logger.error(error);
+			throw new HttpException("Error getting all ratings. ", 500, {cause: error});
+		}
+	}
+
+	public async getPlayersRating(name: string): Promise<Rating | DefaultExceptionDto> {
+		try {
+			const result = await this.ratingRepository.findOne({
+				select: ['name', 'rating', 'ratedgames', 'participation_rating', 'maxrating', 'isbot'],
+				where: { name: name }
+			});
+			if (!result) {
+				throw new HttpException("Error: User not found", 404, {cause: new Error("User not found")});
+			}
+			return result;
+		} catch (error) {
+			this.logger.error(error);
+			throw new HttpException("Error getting players rating. ", 500, {cause: error});
+		}
+	}
+
+	// generates the games and players ratings
+	// the rating calculation is based on Nohatcoders Elo rating system
+	// https://nohatcoder.dk/2015-03-22-1.html
+	// this function gets the games based on the last used game id from the previous.txt file
+	// it gets all the players from the players table
+	// this generateRating function does 3 database transactions
+	// it loops through the games data and updates the players rating for that game and sets if the player has changed
+	// it loops through the players data and updates the players rating for that player if they have changed
+	// it updates the players fatigue rating even if they haven't changed
+	// it tries to commit the transactions and if there is an error it rolls back the transactions
+	// it creates a rating.json file for the ratings.html page
+	// if there is no error it writes the last used game id to the previous.txt file
+	public async generateRating(): Promise<void> {
+		let has_error = false;
+		let previous = '0 0';
+		try {
+			await access(process.env.PREVIOUS_FILE);
+			previous = await readFile(process.env.PREVIOUS_FILE, 'utf-8');
+		} catch (error) {
+			this.logger.error(error);
+			return;
```

Comment:
shouldn't we raise the exception here for any downstream users instead of hiding it in the logs?

Reply:
@nitzel ah re-reading this I see what your referring to; the generate ratings currently isn't directly executed by a end user so throwing an actual error wouldn't do much, we are using new relic which will catch the logged error and can notify it, but it would also do the same for a thrown error so either would work in this.

I'll go ahead and add the throw error in case we do open up a endpoint to manually run the generate ratings